core/src/main/java/org/jdbi/v3/core/mapper/reflect/ImmutableMapper.java

```diff
@@ -0,0 +1,223 @@
+                                     List<String> unmatchedColumns) throws SQLException {
+        final List<RowMapper<?>> mappers = new ArrayList<>();
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+
+        for (PropertyDescriptor descriptor : info.getPropertyDescriptors()) {
+            Nested anno = Stream.of(descriptor.getReadMethod(), descriptor.getWriteMethod())
+                .filter(Objects::nonNull)
+                .map(m -> m.getAnnotation(Nested.class))
+                .filter(Objects::nonNull)
+                .findFirst()
+                .orElse(null);
+
+            if (anno == null) {
+                String paramName = prefix + paramName(descriptor);
+
+                findColumnIndex(paramName, columnNames, columnNameMatchers, () -> debugName(descriptor))
+                    .ifPresent(index -> {
+                        Type type = descriptor.getReadMethod().getGenericReturnType();
+                        ColumnMapper<?> mapper = ctx.findColumnMapperFor(type)
+                            .orElse((r, n, c) -> r.getObject(n));
+
+                        mappers.add(new SingleColumnMapper<>(mapper, index + 1));
+                        properties.add(descriptor);
+
+                        unmatchedColumns.remove(columnNames.get(index));
+                    });
+            } else {
+                String nestedPrefix = prefix + anno.value();
+
+                RowMapper<?> nestedMapper = nestedMappers
+                    .computeIfAbsent(descriptor, d -> new ImmutableMapper(d.getPropertyType(), nestedPrefix))
+                    .specialize0(rs, ctx, columnNames, columnNameMatchers, unmatchedColumns);
+
+                mappers.add(nestedMapper);
+                properties.add(descriptor);
+            }
+        }
+
+        if (mappers.isEmpty() && columnNames.size() > 0) {
+            throw new IllegalArgumentException(String.format("Mapping immutable type %s " +
+                "didn't find any matching columns in result set", type));
+        }
+
+        return (r, c) -> {
+            for (int i = 0; i < mappers.size(); i++) {
+                RowMapper<?> mapper = mappers.get(i);
+                PropertyDescriptor property = properties.get(i);
+
+                Object value = mapper.map(r, ctx);
+
+                try {
+                    property.getWriteMethod().invoke(builder, value);
+                } catch (IllegalAccessException | InvocationTargetException e) {
+                    throw new SQLException();
+                }
+            }
+
+            try {
+                return (T) builder.getClass().getMethod("build").invoke(builder);
+            } catch (IllegalAccessException | NoSuchMethodException e) {
+                throw new IllegalArgumentException();
+            } catch (InvocationTargetException e) {
+                if (e.getTargetException() instanceof IllegalStateException) {
```

Comment:
This catch block looks even stranger.

Reply:
Did it with `MethodHandles` now, but does not get any better, still prefer `MethodHandles` over normal reflection